ENT-696 update course enrollment view for enterprise enabled course modes

zubair-arbi opened the pull request:
@asadiqbal08 @saleem-latif @douglashall 

**Description:** update course enrollment view for enabled course mode in enterprise catalogs

**JIRA:** [ENT-696](https://openedx.atlassian.net/browse/ENT-696)

**Dependencies:** N/A

**Merge deadline:** 20 Oct, 2017

**Installation instructions:** Install `edx-enterprise` from this branch `zub/ENT-696-update-course-enrollment-for-catalog-queryparam` in `edx-platform`.

**Testing instructions:**

1. Create enterprise with a catalog
2. Add a record in table `Enterprise Customer Catalogs` against the above enterprise.
3. Provide desired course mode (in order) to display on enterprise course enrollment page
<img width="973" alt="screen shot 2017-10-13 at 6 07 46 pm" src="https://user-images.githubusercontent.com/5072991/31547815-25cac938-b042-11e7-9e99-ca90fb70359f.png">

4. Now go to enterprise course enrollment page as a learner and verify that the displayed course modes are from the list provided in table `Enterprise Customer Catalogs` against the enterprise.
5. Now update the value for `Enabled course modes:` in enterprise customer catalog record and provide course modes which don't exist for the course.
6.  Go to enterprise course enrollment page as a learner and verify that learner sees a generic info message `Something happened. This course is not available. Please start over and select a different course.` and there are no course modes listed on the page and also there is no `Continue` button.
<img width="829" alt="screen shot 2017-10-13 at 6 39 46 pm" src="https://user-images.githubusercontent.com/5072991/31549129-a5f8cf98-b046-11e7-8361-410afe045e2f.png">


**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [ ] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] Called `make static` for webpack bundling if any static content was updated.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [ ] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.

zubair-arbi commented:
@douglashall Please have a look on the changes and let me know if we are on right track.

douglashall commented:
@zubair-arbi Looks like we are heading in the right direction here. Let me know when you are ready for a final review.

zubair-arbi commented:
@douglashall @saleem-latif this PR is ready for review.

douglashall commented on enterprise/views.py:
```diff
@@ -498,35 +499,68 @@ def set_final_prices(self, modes, request):
             enterprise_customer = get_enterprise_customer_or_404(enterprise_uuid)
-            course, course_run = \
-                CourseCatalogApiServiceClient(enterprise_customer.site).get_course_and_course_run(course_run_id)
         except ImproperlyConfigured:
```
I think you can get rid of the try/catch here now.

douglashall commented on enterprise/views.py:
```diff
@@ -498,35 +499,68 @@ def set_final_prices(self, modes, request):
         into an actual EnterpriseCustomer, a set of details about the course, and a list
         of the available course modes for that course run.
         """
+        enterprise_catalog_uuid = request.GET.get('catalog')
```
Move this down to line 517 closer to where you are using it.

douglashall commented on enterprise/views.py:
```diff
@@ -498,35 +499,68 @@ def set_final_prices(self, modes, request):
+            modes = [mode for mode in modes if mode['slug'] in enterprise_catalog.enabled_course_modes]
+            modes.sort(key=lambda course_mode: enterprise_catalog.enabled_course_modes.index(course_mode['slug']))
+            if not modes:
+                add_generic_info_message_for_error(request)
```
Move 548-563 to a separate function: `get_available_course_modes(course_run_id, enterprise_customer_catalog=enterprise_catalog)`

douglashall commented on enterprise/views.py:
```diff
@@ -498,35 +499,68 @@ def set_final_prices(self, modes, request):
+                    course=course, course_run=course_run, enterprise_uuid=enterprise_customer.uuid
+                )
+            )
             raise Http404
```
Don't raise a 404, display the generic error message.

zubair-arbi replied:
This will be handled as part of [ENT-717] (https://openedx.atlassian.net/browse/ENT-717).

douglashall commented on enterprise/views.py:
```diff
@@ -498,35 +499,68 @@ def set_final_prices(self, modes, request):
             LOGGER.warning('Unable to get course modes for course run id {course_run_id}.'.format(
                 course_run_id=course_run_id
             ))
             raise Http404
```
Don't raise a 404, display the generic error message.

douglashall commented on enterprise/views.py:
```diff
@@ -498,35 +499,68 @@ def set_final_prices(self, modes, request):
+                    enterprise_customer.site
+                ).get_course_and_course_run(course_run_id)
+            except ImproperlyConfigured:
+                raise Http404
```
If the `catalog` query parameter was provided and we can't get the course and course_run data for any reason, whether that be because there was a problem retrieving the `EnterpriseCustomerCatalog` for the UUID provided or because the course run does not exist in the specified catalog, we should should display the landing page with an error message, the course details (if we are able to retrieve them), and NO mode options or continue button. It seem like we should be able to simplify the code here looking at things from that approach.

zubair-arbi replied:
Replacing `raise Http404` with `add_generic_info_message_for_error(request)` remaining two places is causing multiple tests to fail. We should create a separate story for remaining refactoring.

douglashall replied:
Sounds good. Can you create those tickets and tag me?

UmanShahzad replied:
I think the existing https://openedx.atlassian.net/browse/ENT-688 would be enough?

zubair-arbi replied:
@douglashall @UmanShahzad I have created a separate ticket [ENT-717](https://openedx.atlassian.net/browse/ENT-717) just in case.

UmanShahzad commented on enterprise/messages.py:
```diff
@@ -116,3 +116,26 @@ def add_not_one_click_purchasable_message(request, enterprise_customer, program_
     )
+
+
+def add_generic_info_message_for_error(request):
```
Yep, I think mine will get merged soon as it's a smaller / simpler PR, so you can reuse that and just pass in `'course'` to the 2nd parameter.

zubair-arbi replied:
I have asked Ian and Christine about this on jira ticket.

UmanShahzad replied:
This is something we did confirm, and #219 is now merged with this message. So after a rebase you can use that function I've created.

douglashall replied:
@zubair-arbi Okay, it looks like UX wants a different message in this scenario. Just rename this function to `add_generic_error_message`.

zubair-arbi replied:
@douglashall since this function will add an INFO message for any error on server side that's why I named it `add_generic_info_message_for_error`.

UmanShahzad commented on enterprise/models.py:
```diff
@@ -961,6 +961,22 @@ def get_course_run(self, course_run_id):
+        if not self.contains_content_items('key', [course_run_id]):
+            return None, None
+
+        return CourseCatalogApiServiceClient(self.enterprise_customer.site).get_course_and_course_run(course_run_id)
```
In case this needs to be handled gracefully, would need a try-catch block for `ImproperlyConfigured`.

zubair-arbi replied:
We don't need to hide any error related to configuration. I will add `Raises` section in the docstring about the exception `ImproperlyConfigured`.

zubair-arbi replied:
Done

zubair-arbi commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -30,47 +30,55 @@ <h2 class="course-confirmation-title">{{ confirmation_text }}</h2>
+            </div>
+            {{ course_short_description }}
+            {{ view_course_details_text|link_to_modal:0 }}
+          </div>
```
In the mockup which Ian shared on jira ticket ENT-696 there was only course image, course name and organization name. Please confirm if we want to display the other course details too.

douglashall replied:
Okay, let's go with what Ian has.

zubair-arbi commented:
@douglashall @UmanShahzad Please have another look.

douglashall commented on enterprise/views.py:
```diff
@@ -498,36 +499,70 @@ def set_final_prices(self, modes, request):
-                course=course, course_run=course_run
-            ))
-            raise Http404
+        enterprise_catalog_uuid = request.GET.get('catalog')
```
Add this comment above this line:
```
# If the catalog query parameter was provided, we need to scope
# this request to the specified EnterpriseCustomerCatalog.
```

douglashall commented on enterprise/views.py:
```diff
@@ -498,36 +499,70 @@ def set_final_prices(self, modes, request):
+                raise Http404
+
+        if not course or not course_run:
+            LOGGER.warning(
```
Add comment above this line:
```
# The specified course either does not exist in the
# specified EnterpriseCustomerCatalog, or does
# not exist at all in the discovery service.
```

douglashall commented on enterprise/views.py:
```diff
@@ -498,36 +499,70 @@ def set_final_prices(self, modes, request):
 
         course_modes = []
+        if enterprise_catalog_uuid and not enterprise_catalog:
+            return enterprise_customer, course, course_run, course_modes
```
Add comment above this line:
```
# A catalog query parameter was given, but the
# specified EnterpriseCustomerCatalog does not
# exist, so just return and display the generic
# error message.
```

douglashall commented on enterprise/views.py:
```diff
@@ -498,36 +499,70 @@ def set_final_prices(self, modes, request):
-    def get_base_details(self, enterprise_uuid, course_run_id):
+    def get_available_course_modes(self, request, course_run_id, enterprise_catalog):
+        """
+        Filter and order available course modes.
```
```
"""
Return the available course modes for the course run.

The provided EnterpriseCustomerCatalog is used to filter and order the course
modes returned using the EnterpriseCustomerCatalog's enabled_course_modes
field.
"""
```

zubair-arbi commented:
Rebased onto master and squashed from 65053687c068b6d39d29adee989d201237cad308.
@douglashall I add added comments according to your feedback.
@UmanShahzad @saleem-latif  Please review.

UmanShahzad commented:
Thanks @zubair-arbi ! I didn't do a significant enough review so I'm going to skimp on giving a :+1: as it wouldn't be that meaningful from me.

zubair-arbi commented:
I have verified changes on business sandbox.

asadiqbal08 commented on enterprise/messages.py:
```diff
@@ -101,3 +101,26 @@ def add_unenrollable_item_message(request, item):
+    """
+    messages.info(
+        request,
+        _(
```
Have you added the corresponding translations?

zubair-arbi replied:
Nope. @saleem-latif will add the translations in his [PR](https://github.com/edx/edx-enterprise/pull/217).
